File: flutter_custom_tabs_android/android/src/main/java/com/github/droibit/flutter/plugins/customtabs/CustomTabsPlugin.java
package com.github.droibit.flutter.plugins.customtabs;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;

public class CustomTabsPlugin implements FlutterPlugin, ActivityAware {
    private @Nullable CustomTabsLauncher api;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        api = new CustomTabsLauncher();
        Messages.CustomTabsApi.setUp(binding.getBinaryMessenger(), api);
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPlugin.FlutterPluginBinding binding) {
        if (api == null) {
            return;
        }

        Messages.CustomTabsApi.setUp(binding.getBinaryMessenger(), null);
        api = null;
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        if (api == null) {
            return;
        }
        api.setActivity(binding.getActivity());
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity();
    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
        onAttachedToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        if (api == null) {
            return;
        }
        api.setActivity(null);
    }
}

File: flutter_custom_tabs_android/android/src/main/kotlin/com/github/droibit/flutter/plugins/customtabs/CustomTabsPlugin.kt
package com.github.droibit.flutter.plugins.customtabs

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

class CustomTabsPlugin : FlutterPlugin, ActivityAware {
    private var api: CustomTabsLauncher? = null

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        api = CustomTabsLauncher()
        CustomTabsApi.setUp(binding.binaryMessenger, api)
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        if (api == null) {
            return
        }

        CustomTabsApi.setUp(binding.binaryMessenger, null)
        api = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        val api = this.api ?: return
        api.setActivity(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        val api = this.api ?: return
        api.setActivity(null)
    }
}
